### PR TITLE
Horreum Reporter

### PR DESCRIPTION
**Pull request by diegolovison**

### CODE REVIEW

File: reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            if (computeTotal) {
+               if (aggregated == null)
+                  aggregated = nodeSummary.copy();
+               else
+                  aggregated.merge(nodeSummary);
+            }
+            firstTimestamp = Math.min(firstTimestamp, nodeSummary.getBegin());
+            lastTimestamp = Math.max(lastTimestamp, nodeSummary.getEnd());
+         }
+         if (computeTotal && aggregated != null) {
```
Comment (gustavolira) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java:63]: You can remove this if and add this logic into line 58

File: reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      File parentDir = new File(targetDir);
+      if (!parentDir.exists()) {
+         if (!parentDir.mkdirs()) {
+            log.warn("Issues creating parent dir " + parentDir);
+         }
+      } else {
+         if (!parentDir.isDirectory()) {
+            throw new IllegalStateException(targetDir + " is not a directory");
+         }
+      }
```
Comment (gustavolira) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java:152]: I think you can replace those lines with
```
try {
            Files.createDirectory(file.toPath());
        } catch (IOException e) {
            throw new IllegalStateException("Issues creating parent dir " + parentDir);
        }
 ```
 The Files.createDirectories() creates a new directory and parent directories that do not exist. This method does not throw an exception if the directory already exists.

File: reporters/reporter-horreum/src/main/java/org/radargun/reporting/horreum/HorreumReporter.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+      List<Map<String, Double>> convertedValues = new ArrayList<>();
+      for (Map<String, String> map : values) {
+         Map<String, Double> newMap = new HashMap<>();
+         for (String key : map.keySet()) {
+            newMap.put(key, Double.valueOf(map.get(key)));
+         }
+         convertedValues.add(newMap);
+      }
+      return convertedValues;
+   }
```
Comment (gustavolira) [inline reporters/reporter-horreum/src/main/java/org/radargun/reporting/horreum/HorreumReporter.java:195]: WDYT to use
```
private List<Map<String, Double>> convertData2(List<Map<String, String>> values) {
       return values.stream().map(it -> it.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, e -> Double.valueOf(e.getValue()))))
                .collect(Collectors.toList());
    }
 ```

File: reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      columns.addAll(rowData.keySet());
+
+      rowData.put(WORKER_INDEX, String.valueOf(workerStats.getKey()));
+      rowData.put(ITERATION, String.valueOf(it));
+      rowData.put(PERIOD, String.valueOf(summary.getEnd() - summary.getBegin()));
+      rowData.put(THREAD_COUNT, String.valueOf(workerStats.getValue().size()));
+      return summary;
+   }
+
+   private static void addRepresentations(Statistics summary, Map<String, String> rowData, Statistics statistics,
```
Comment (gustavolira) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java:100]: `Parameter 'summary' is never used `

Comment (gustavolira) [reply reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java:100]: WDYT to use:
```
private static void addRepresentations(Statistics summary, Map<String, String> rowData, Statistics statistics,
                                          String operationName, double[] percentiles) {
      addDefaultOutcome(summary, rowData, statistics, operationName, percentiles);
      addOperationThroughput(summary, rowData, statistics, operationName, percentiles);
      addDataThroughput(summary, rowData, statistics, operationName, percentiles);
      addPercentiles(summary, rowData, statistics, operationName, percentiles);
      addMeanAndDev(summary, rowData, statistics, operationName, percentiles);
   }

   private static void addMeanAndDev(Statistics summary, Map<String, String> rowData, Statistics statistics,
                                     String operationName, double[] percentiles) {
      MeanAndDev meanAndDev = statistics.getRepresentation(operationName, MeanAndDev.class);
      if (meanAndDev != null) {
         rowData.put(operationName + ".Mean", String.valueOf(round(meanAndDev.mean, 2)));
         rowData.put(operationName + ".Dev", String.valueOf(round(meanAndDev.dev, 2)));
      }
   }

   private static void addPercentiles(Statistics summary, Map<String, String> rowData, Statistics statistics,
                                      String operationName, double[] percentiles) {
      Percentile[] percentileValues = statistics.getRepresentation(operationName, Percentile[].class);
      if (percentileValues != null) {
         for (int i = 0; i < percentiles.length; i++) {
            rowData.put(operationName + ".Percentile" + (int) (percentiles[i] * 100), String.valueOf(percentileValues[i].value));
         }
      }
   }

   private static void addDataThroughput(Statistics summary, Map<String, String> rowData, Statistics statistics,
                                         String operationName, double[] percentiles) {
      DataThroughput dataThroughput = statistics.getRepresentation(operationName, DataThroughput.class);
      if (dataThroughput != null) {
         rowData.put(operationName + ".DataThroughput", String.valueOf(round(dataThroughput.net, 1)));
         rowData.put(operationName + ".DataThroughputWithErrors", String.valueOf(round(dataThroughput.gross, 1)));
      }
   }

   private static void addOperationThroughput(Statistics summary, Map<String, String> rowData, Statistics statistics,
                                              String operationName, double[] percentiles) {
      OperationThroughput throughput = statistics.getRepresentation(operationName, OperationThroughput.class);
      if (throughput != null) {
         rowData.put(operationName + ".Throughput

File: reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java (hunk 2 above)
Comment (diegolovison) [reply reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java:152]: The content of `DataReporter` comes from `CsvReporter`. How can I add the warning and throw the exception if the `parentDir` is a file using `Files.createDirectory`?

File: reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java (hunk 1 above)
Comment (diegolovison) [reply reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java:63]: Done.

**Comment by diegolovison** [comment]
Those changes are nice to have. Could we do it in another PR? Basically, the content of the class `DataReporter` came from `CsvReporter`. Changing the content of the class means that I need to retest CSV and Horreum again.

File: reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+import org.radargun.reporting.Report;
+import org.radargun.stats.Statistics;
+import org.radargun.stats.representation.DataThroughput;
+import org.radargun.stats.representation.DefaultOutcome;
+import org.radargun.stats.representation.MeanAndDev;
+import org.radargun.stats.representation.OperationThroughput;
+import org.radargun.stats.representation.Percentile;
+import org.radargun.utils.Utils;
+
+public class DataReporter {
```
Comment (alanfx) [inline reporters/reporter-default/src/main/java/org/radargun/reporting/commons/DataReporter.java:27]: Some class JavaDoc here would be good! I can see this class is used by the `CsvReporter` class below, so there must be something good this class is doing! :-)

File: reporters/reporter-horreum/README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
+PerfRepo reporter
```
Comment (alanfx) [inline reporters/reporter-horreum/README.md:1]: Guessing this should be "Horreum" instead?

**Review by alanfx — CHANGES_REQUESTED** [review]
Hey @diegolovison, this looks good to me. Just a few minor comments

**Pull request merged**